### PR TITLE
mapState: allow Partial<TState> return

### PR DESCRIPTION
**Pull request by kLabz**
## Changes
`mapState` can now return either `TState` or `Partial<TState>`, in order to allow components to mix local state and data from the store.

### CODE REVIEW

File: redux/react/ConnectMacro.hx [hunk 1]
```diff
@@ -200,12 +279,29 @@ class ConnectMacro
 		});
 	}
 
-	static function addCache(fields:Array<Field>)
+	static function addCache(fields:Array<Field>, inClass:ClassType, mapStateIsPartial:Bool)
 	{
+		var mapStateType:ComplexType = macro :Dynamic;
+
+		if (mapStateIsPartial)
```
Comment (elsassph) [inline redux/react/ConnectMacro.hx]: Do we care about not defining `__state` as `Dynamic`?

File: redux/react/ConnectMacro.hx [hunk 2]
```diff
@@ -176,7 +255,7 @@ class ConnectMacro
 		return false;
 	}
 
-	static function addConnect(fields:Array<Field>)
+	static function addConnect(fields:Array<Field>, mapStateIsPartial:Bool)
```
Comment (elsassph) [inline redux/react/ConnectMacro.hx]: `mapStateIsPartial` isn't used

File: redux/react/ConnectMacro.hx [hunk 3]
```diff
@@ -220,17 +316,16 @@ class ConnectMacro
...
 	{
 		var propsArg = { name: 'props', type: macro :Dynamic };
 		var contextArg = { name: 'context', type: macro :Dynamic };
 
-		var updateStateExpr = macro {
-			__state = mapState(context.store.getState(), props);
+		var initCacheExpr = macro __state = mapState(context.store.getState(), props);
 
-			if (__state != null)
-				state = react.ReactUtil.copy(state == null ? {} : state, __state);
```
Comment (elsassph) [inline redux/react/ConnectMacro.hx]: Not trying to merge with a potentially already user-defined `state` value?

Comment (kLabz) [reply redux/react/ConnectMacro.hx]: Not here. It is handled below (lines 338 ~ 350)

File: redux/react/ConnectMacro.hx (hunk 2 above)
Comment (kLabz) [reply redux/react/ConnectMacro.hx]: Removed

File: redux/react/ConnectMacro.hx (hunk 3 above)
Comment (kLabz) [reply redux/react/ConnectMacro.hx]: However, states are _not_ merged if mapState returns a "complete" state. Should they be merged?

File: redux/react/ConnectMacro.hx (hunk 1 above)
Comment (kLabz) [reply redux/react/ConnectMacro.hx]: Not really

File: redux/react/ConnectMacro.hx [hunk 4]
```diff
@@ -240,6 +237,17 @@ class ConnectMacro
 					case FFun(f):
 						if (f.args.length < 1) f.args.push(propsArg);
 						if (f.args.length < 2) f.args.push(contextArg);
+
+						var updateStateExpr = macro {
+							if (state == null)
```
Comment (elsassph) [inline redux/react/ConnectMacro.hx]: It looks like we're generating redundant `__state = mapState(context.store.getState(), props)` compiled in the resulting JS.

**Pull request merged**